README.md
Minitcp
===

Presentation
==

A little tool for doing some TCP/UDP sockets communications.

This tool have no pretention : it is not EventMachine ! : 

* no mainloop, 
* many threads are created/deleted.
* it is useful for send some data one-shot, test a ascii communication in few minutes...
* should be equivalent to netcat+bash ...


A  TCP client :

```ruby
MClient.run_one_shot("localhost",2200) do |socket|
   socket.on_any_receive { |data| p "client recieved #{data.inspect}"}
   3.times { |j| socket.puts "Hello  #{j}..." ; sleep(1) }
end.join
```

An echo server :

```ruby
srv=MServer.service(2200,"0.0.0.0",22) do |socket|
  socket.on_any_receive { |data|  socket.print(data) }
  socket.on_timer(2000) do
    socket.close
  end
  socket.wait_end
end
```

A UDP server :

```ruby
	UDPAgent.on_datagramme("127.0.0.2",SRV_PORT ) { |data,from,p| 
	  puts "Agent: received #{data} from #{from}:#{p}" 
	  data && data.size>3 ? "OK-#{data}." : nil
	}
```

A UDP sender :

```ruby
	UDPAgent.on_timer(1000, 
		port: 2232,
		on_timer: proc do
		  data=Time.now.to_i.to_s
		  puts "\n\n\non timer send <#{data}>"
		  {mess: data,host: "127.0.0.2",port: SRV_PORT}
		end,
		on_receive: proc { |data,from,sock| 
		  puts "Client: received #{data} from #{from}"  
		  UDPAgent.send_datagram_on_socket(sock,from.last,from[1],'ack')
		}
	)
```

A Message agent :
Message Client:
```ruby
MClientAgent.run("localhost",2222) do |chan|
  chan.send_message({date: Time.now.to_f}) 
  chan.on_message do |mess| 
	puts "cli: receive: #{mess.inspect}" 
	chan.close
	nil
  end
  chan.on_timeout(10_000) { p "timeout" ; chan.close rescue nil }
  chan.wait_end
end
```

Message Server:
```ruby
MServerAgent.run(2222,"localhost",22) do |chan|
    chan.on_message { |mess|  p mess ; "ok" }
    chan.wait_end
end
```

Docs: http://rubydoc.info/gems/minitcp


TCP
===

Client
---

```
MClient.run_one_shot(host,port) do |socket| ... end
MClient.run_continious(host,port,time_inter_connection) do |socket| ... end
```

Client socket is extended with all specifiques commandes (```SocketReactive```, see Sockets).

Server
---

```ruby
srv=MServer.service(port,"0.0.0.0",max_client) do |socket| ... end
```
Mserver run a GServer. 
Connected sockets are extended with same module as in client.

Sockets
---

Handlers are disponibles for any sockets : client or server. 
All handler's bloc run in distinct thread : so any handler-bloc can wait anything.
if socket is closed, handler/thread are cleanly (?) stoped.

* socket.**on_any_receive() {|data| ...}**          : on receive some data, any size
* socket.**on_n_receive(sizemax=1) {|data| ...}**   : receives n byte(s) only
* socket.**on_receive_sep(";") {|field | ... }**    : reveive data until string separator
* socket.**on_timer(value_ms) { ... }**             : each time do something, if socket is open
* socket.**after(duration) { ... }**    : do something after n millisecondes, if socket is open

Some primitives are here for help (no thread):

* **received_timeout(sizemax,timeout)** : wait for n bytes, with timeout, (blocking caller),
* **wait_end()**                        : wait, (blocking caller) until socket is close. this
  work only if something has closed the socket.this is possible unicly by receiving 0 bte on a
* **receive_n_bytes** / **on_receive_sep** : bocking version of handler,
* **connected?** : test if a close have been done,  
* **data_readed()** : some receives (receive_sep...) can read more data form sockets,
 this data are used by ```n_revieve/any_receive/receive_sep```, but they can be read/reseted 
 whith data_readed getseter.

This primitives are declared in SocketReactive module.

UDP
===

2 type of agents : 

* **datagramme** (server) : 
   receive data from anywhere, can reply to sender : UDPAgent.on_datagramme
* **Timer**  (client) : 
   can emit to everybody, periodicly. Can receive response from them : UDPAgent.on_timer
   ```send_datagram_on_socket``` can be used for emit spontaneously to a server

2 primitives :

* **send_datagram(host,port,message)** : 
   create a socket, send mesage and close socket (can't receive a reply)
* **send_datagram_on_socket(socket,host,port,message)** : 
  use an existant socket for send a message to ip:port
  the existant socket should be one created by a current UDPAgent timer or server

Messages
===
Serveur and client ```MServerAgent``` and ```MClientAgent``` define a 'agent ' which
can communicate by messages.

Messages are ```ruby-data.inspect```, max size is 1 MByte.
* MClientAgent.run(host,port) : connect to a MServerAgent, for communicate by message
* MServerAgent.run(port,host,max-connections) : message server

in Agent, socket are extended with this capability :
* socket.send_message(data)
* socket.on_message { |data|  .... ; ret_data}  will send ret_data to sender if not nil

See test.rb for an example

TODO
==

* Serial line
* more socket primitive
* messages Agent : use Marchal instead of inspect/eval ....

Tests case
==
A TCP proxy, debug tool (see samples/proxy.rb) :

```ruby
MServer.service(2200,"0.0.0.0",22) do |scli|
  px 2, "======== client Connected ========"
  srv=MClient.run_one_shot("ip",2200) do |ssrv|
     px 1, "======== server Connected ========"
     ssrv.on_any_receive { |data| px 1,data; scli.print data }
     scli.on_any_receive { |data| px 2,data; ssrv.print data}
     ssrv.wait_end
     scli.close rescue nil
  end
  scli.wait_end
  srv.stop rescue nil
end   
def px(sens,data)
  data.each_line {|line| puts "#{(sens==1 ? "> " : "< ")}#{line.chomp}"
end

```




Test serial **protocole-like** : header/body => ack/timeout:

* client send <length><data> , wait one char acquit or timeout
* serveur receive heade( size: 4 chars) , then data with size, send ack


```ruby
   
srv=MServer.service(2200,"0.0.0.0",22) do |socket|
  socket.on_n_receive(4) { |data| 
     size=data.to_i
     data=socket.recv(size)
     puts "  Server recieved buffer : #{data.inspect}"
     if rand(100)>50
        socket.print("o") 
     else 
        puts "  !!!non-ack by serv"
     end
  }
  socket.on_timer(40*1000) { puts " serv close after 40 seconds"; socket.close }
  socket.wait_end
end   

MClient.run_one_shot("localhost",2200) do |socket|
   10.times { |j| 
	   size=rand(1..10)
	   puts "Sending #{size} data..."
	   data='*'*size
	   socket.print "%04d" % size
	   socket.print data 
	   p socket.received_timeout(1,100)  ? "ack ok" : "!!! timeout ack"
   }
   p "end client"
end.join


```



CHANGELOG.txt
0.17.1
    minitcp.rb : updage comments on UDPAgent
    README.md : update udp docs
0.17.0 : 2016-05-18 16:00:30 +0200
0.16.0 : 2016-05-18 16:00:15 +0200
    minitcp.rb : correction UDPAgent
    README.md : update
0.15.0 : 2015-04-02 15:51:14 +0200
    minitcp.rb : force data received to encding BINARY
0.14.0 : 2015-01-13 20:56:17 +0100
    minitcp.rb : no audit, no log for tcp server
    relaiproxy.rb : notraces
    relai_config.data : test1 aded
    minitcp.rb : no traces
0.13.0 : 2014-11-28 13:35:33 +0100
    plot.rb : window resize enable
    pingpongplot.rb : multiple url test
    README.md : message agent expample
    README.md : massage agent expalple
    README.md : example message
    README.md : example for agent message
0.12.0 : 2014-10-30 21:31:33 +0100
0.11.0 : 2014-10-30 21:30:28 +0100
    README.md : ocpp
    README.md : plugin doc
    Rakefile.rb : version management
    README.md : plugin ocpp description
0.10.0 : 2014-09-24 20:18:54 +0200
    minitcp.rb : reveive_n_byte correction
0.9.0 : 2014-09-18 22:38:31 +0200
0.7.0 : 2014-09-18 22:37:45 +0200
    README.md : titles
    test.rb : UDP tests
    minitcp.rb : udp creation
    README.md : UDP creation
    statcpu.rb : windows version, not finish...
    README.md : spdy start demo
    README.md : best code example
    README.md : presentation
    README.md : continious
0.6.1
    plot.rb : refactoring Measure.class
    pingpongplot.rb : temporisation
0.6.0 : 2014-08-21 11:06:44 +0200
    README.md : todo
    plot.rb : auto-scale
    name_server.rb : creation
    plot.rb : color
    plot.rb : interaction with mouse
    name_server.rb : forget debug
0.5.0 : 2014-08-18 16:39:35 +0200
0.4.0 : 2014-08-18 16:37:12 +0200
    minitcp.rb : error on stdout for SocketReactive
    README.md : typo
    README.md : sockets presentation
    README.md : rdoc url
    minitcp.rb : comments
    README.md : presentations, format
    proxy.rb : prefix correction
    README.md : proxy correction
    README.md : todo, presentation
0.3.0 : 2014-08-18 01:27:40 +0200
    test.rb : sleep before thread good stoping detection
    minitcp.rb : all read hendler do socket close when necessary
    README.md : todo
0.2.0 : 2014-08-17 16:07:03 +0200
    Rakefile.rb : update
    test.rb : license
    proxy.rb : license

samples/name_server.rb
# LGPL,  Author: Regis d'Aubarede <regis.aubarede@gmail.com>
#################################################################
#  name_server.rb
#################################################################
require_relative '../lib/minitcp.rb'

BasicSocket.do_not_reverse_lookup = true
Thread.abort_on_exception = true

$SIZE=30
$dico_name={}
$dico_ip={}

def run_net_relay()
  MServer.service(4410,"0.0.0.0",1000) do |socket| 
    socket.on_n_receive($SIZE) do |data|
      ip=socket.remote_address.ip_address
      cmd,url,http=data.split(' ')  
      name,method,*args=url.split('/').last
      next if cmd!="GET"  
      case method
      when "server"
        old=ActiveConnection.search(name)
        if old && old.state==:free
           old.deconnect
           old=nil
        end
        if !old || old.state==:connected
          ActiveConnection.add(socket)
        end
      when "client"
        if srv=ActiveConnection.search(args.first)
          srv.connect_to(name)
        else
          smess("NOTCONNECT")
        end
      end
      rep=""
    end
  end
end

def smess(socket,*args)
    data=  args.join(";")+";"
    socket.send(data+" "*($SIZE-data.length),0) rescue p $!
end

def run_server_name()
  MServer.service(4400,"0.0.0.0",100) do |socket| 
    socket.on_n_receive($SIZE) do |data|
      ip=socket.remote_address.ip_address
      cmd,*args=data.split(";")  
      name=args.first
      rep=""
      case cmd
        when "update"
          $dico_name[name]=ip
          $dico_ip[ip]=name
          rep="OK"
        when "forget"
          oldip=$dico_ip[ip]
          if oldip && $dico_ip[ip]==name && $dico_name[name]==ip
            $dico_name.delete(name)
            $dico_ip.delete(ip)
            rep="OK"
          end
        when "get"
          rep=$dico_name[name].to_s
        when "geti"
          rep=$dico_ip[name].to_s
      end
      rep.size>0  ? smess(socket,"ACK",name,rep) : smess(socket,"NACK","NOK","NOK")
    end
    socket.after(10*1000) { socket.close rescue nil}
    socket.wait_end
  end
end

################# Client api ########################

def update_server_name(hostname)
  MClient.run_one_shot("localhost",4400) do |socket|
    socket.on_n_receive($SIZE) { |data| p data ; socket.close }
    smess(socket,"update",hostname)
    socket.wait_end
  end
end
def fetch_name(hostname)
  ret=""
  MClient.run_one_shot("localhost",4400) do |socket|
    socket.on_n_receive($SIZE) { |data| ret= data.split(";")[2] ; socket.close  }
    smess(socket,"get",hostname)
    socket.wait_end
    ret
  end.join
  ret
end
def fetch_ip(ip)
  ret=""
  MClient.run_one_shot("localhost",4400) do |socket|
    socket.on_n_receive($SIZE) { |data| ret= data.split(";")[2] ; socket.close  }
    smess(socket,"geti",ip)
    socket.wait_end
    ret
  end.join
  ret
end
def forget_me(name)
  MClient.run_one_shot("localhost",4400) do |socket|
    ret="?"
    socket.on_n_receive($SIZE) { |data| ret= data.split(";")[2] ; socket.close  }
    smess(socket,"forget",name)
    socket.wait_end
    p "forget : #{ret}"
  end.join
end

if $0==__FILE__
  run_server_name
  
  Thread.new { 4.times { update_server_name("glurp") ; sleep 1 } }
  Thread.new { 4.times { update_server_name("glurp22") ; sleep 1 } }

  sleep 2
  puts "\n\n***************** get name *******************\n\n"

  p fetch_name("glurp")
  p fetch_ip("127.0.0.1")
  forget_me("glurp22")
  sleep
end
